### ISSUE TITLE
LS device, attempt to subtract with overflow assert failure

### ISSUE THREAD

**Issue by clemej** [Nortank12/deepcool-digital-linux#19]
Ran into this while testing out the program on my system:
 
Ryzen 5950x CPU
6.12.0 kernel

```
RUST_BACKTRACE=1 ./deepcool-digital-linux -m auto
Device found: LS-SE-DIGITAL
-----
DISP. MODE: auto
TEMP. UNIT: ˚C ˚F
ALARM:      off
-----
Update interval: 750ms

Press Ctrl+C to terminate
thread 'main' panicked at src/monitor/cpu.rs:49:28:
attempt to subtract with overflow
stack backtrace:
   0: rust_begin_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
   2: core::panicking::panic
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:127:5
   3: deepcool_digital_linux::monitor::cpu::Cpu::get_power
             at /oldroot/home/john/git/deepcool-digital-linux/src/monitor/cpu.rs:49:28
   4: deepcool_digital_linux::devices::ls_series::Display::status_message
             at /oldroot/home/john/git/deepcool-digital-linux/src/devices/ls_series.rs:89:29
   5: deepcool_digital_linux::devices::ls_series::Display::run
             at /oldroot/home/john/git/deepcool-digital-linux/src/devices/ls_series.rs:52:35
   6: deepcool_digital_linux::main
             at /oldroot/home/john/git/deepcool-digital-linux/src/main.rs:102:13
   7: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

It runs for a little while (several minutes) before this triggers.  I'm aware I can run in release mode and not see this assert/crash. 

I'm testing a few obvious fixes, such as converting the types to f64 before the subtraction... but I guess the only way this can happen is if the AMD RAPL sysfs file produces output that's not monotonically increasing (or the rounding in this code is off)?

**Comment by Nortank12** [comment]
Try to observe the output of the RAPL file
```bash
watch -n 1 cat /sys/devices/virtual/powercap/amd-rapl/...
```
You can also make a bash script to log the subtracted results.

I haven't tried AMD RAPL so I don't know how it works exactly. The error you sent me is possibly caused by a negative result.

**Comment by clemej** [comment]
It looks like the `energy_uj` counter wraps approximately every 20-30 minutes on my system.  Digging a little deeper, at least some AMD system, like my AM4 system use a 32-bit counter.   The highest values I saw seem to correspond to about `2^32 * 15.3`, or ~`65712999628`  . This aligns with this out of tree kernel amd power driver from https://github.com/amd/amd_energy that has the quote:

```
Energy information (in Joules) is based on the multiplier, 1/2^ESU; where ESU is an unsigned integer read from MSR_RAPL_POWER_UNIT register. Default value is 10000b, indicating energy status unit is 15.3 micro-Joules increment.
```

Short answer.. I watched it approach that number, and them is immediately jumped back down. That would do it.  

For the record, I changed the code to cast the u64's to f64's before the subtraction and it's been running overnight and all day today no issues, although I'm sure I'm getting a glitchy measurement for 1 second every ~30 minutes. 

If I get the inspiration, I'll add some code to handle the counter wrapping. 

Thanks for the program!

**Comment by clemej** [comment]
In fact, it looks like the maximum value is also in sysfs: 

```
$ sudo cat /sys/devices/virtual/powercap/intel-rapl/intel-rapl\:0/max_energy_range_uj
65532610987
```

**Comment by clemej** [comment]
Here's a quick patch that uses the above `max_energy_range_uj` file and respects wrapping on my AMD system. Feel free to use or ignore, I have no idea if it'll break existing Intel systems.  Thanks again for the program! 

```diff
diff --git a/src/monitor/cpu.rs b/src/monitor/cpu.rs
index 0f9cf45..b6194aa 100644
--- a/src/monitor/cpu.rs
+++ b/src/monitor/cpu.rs
@@ -1,17 +1,20 @@
 //! Reads live CPU data from the Linux kernel.

 use crate::error;
+use crate::warning;
 use cpu_monitor::CpuInstant;
 use std::{fs::read_dir, fs::read_to_string, process::exit};

 pub struct Cpu {
     temp_sensor: String,
+    max_uj: u64,
 }

 impl Cpu {
     pub fn new() -> Self {
         Cpu {
             temp_sensor: find_temp_sensor(),
+            max_uj: read_max_energy(),
         }
     }

@@ -46,7 +49,12 @@ impl Cpu {
     ///
     /// Formula: `W = ΔμJ / (Δms * 1000)`
     pub fn get_power(&self, initial_energy: u64, delta_millisec: u64) -> u16 {
-        let delta_energy = self.read_energy() - initial_energy;
+        let current_energy: u64 = self.read_energy();
+        let delta_energy: u64 = if current_energy >= initial_energy {
+            current_energy - initial_energy
+        } else {
+            (self.max_uj - initial_energy) + current_energy
+        };

         (delta_energy as f64 / (delta_millisec * 1000) as f64).round() as u16
     }
@@ -88,3 +96,12 @@ fn find_temp_sensor() -> String {
     error!("Failed to locate CPU temperature sensor");
     exit(1);
 }
+
+fn read_max_energy() -> u64 {
+    let data = read_to_string("/sys/class/powercap/intel-rapl/intel-rapl:0/max_energy_range_uj").unwrap_or_else(|_| {
+        warning!("Failed to get max energy file, using 2^64-1");
+        u64::MAX.to_string()
+    });
+
+    data.trim_end().parse::<u64>().unwrap()
+}
```

**Comment by Nortank12** [comment]
I was able to check 2 CPUs and they have different limits according to `max_energy_range_uj`.
- AMD Ryzen 7 5700X: `65532610987`
- Intel Core i7-8750H: `262143328850`

I used the code you provided and made some modifications to the error handling. Since not all devices require power consumption information, I simply return 0 if the file is not found. However, on the LS and LD series if the value is 0, the program will exit before writing any data since power readings will be inconsistent.

> I have no idea if it'll break existing Intel systems.

It should help on Intel systems as well since the energy reading cannot increase infinitely. The only advantage (based on my test) was that the limit is 4x larger so it has a smaller chance to crash the program.

I uploaded a new release and it should not crash anymore.

**Thank you for the insightful investigation and your contribution!**
